src/app-layout/visual-refresh-toolbar/index.tsx

```diff
@@ -1,563 +1,91 @@
-      if (activeDrawer && drawers?.find(drawer => drawer.id === drawerId)) {
-        return;
-      }
-      // get the size of drawerId. it could be either local or global drawer
-      const combinedDrawers = [...(drawers || []), ...globalDrawers];
-      const newDrawer = combinedDrawers.find(drawer => drawer.id === drawerId);
-      if (!newDrawer) {
-        return;
-      }
-      const newDrawerSize = Math.min(
-        newDrawer.defaultSize ?? drawerSizes[drawerId] ?? MIN_DRAWER_SIZE,
-        MIN_DRAWER_SIZE
-      );
-      //   check if the active drawers could be resized to fit the new drawers
-      //   to do this, we need to take all active drawers, sum up their min sizes, truncate it from resizableSpaceAvailable
-      //   and compare a given number with the new drawer id min size
-
-      // the total size of all global drawers resized to their min size
-      const availableSpaceForNewDrawer = resizableSpaceAvailable - totalActiveDrawersMinSize;
-      if (availableSpaceForNewDrawer >= newDrawerSize) {
-        return;
-      }
-
-      // now we made sure we cannot accommodate the new drawer with existing ones
-      closeFirstDrawer();
+import { SkeletonSlotsAttributes } from './skeleton/interfaces';
+import { DeduplicationType, useMultiAppLayout } from './skeleton/multi-layout';
+import { StateManager } from './state';
+import { getPropsToMerge, mergeProps } from './state/props-merger';
+import { ToolbarProps } from './toolbar';
+
+const AppLayoutStateProvider: React.FC<{
+  forceDeduplicationType: DeduplicationType;
+  appLayoutProps: AppLayoutInternalProps;
+  stateManager: React.MutableRefObject<StateManager>;
+  children: (
+    registered: boolean,
+    appLayoutState: AppLayoutPendingState,
+    toolbarProps: ToolbarProps | null,
+    skeletonAttributes: SkeletonSlotsAttributes
+  ) => React.ReactNode;
+}> = ({ forceDeduplicationType, appLayoutProps, stateManager, children }) => {
+  const [appLayoutState, setAppLayoutState] = useState<AppLayoutPendingState>({ isIntersecting: true });
+  const [skeletonAttributes, setSkeletonAttributes] = useState<SkeletonSlotsAttributes>({});
+  // use { fn: } object wrapper to avoid confusion with callback form of setState
+  const [deduplicator, setDeduplicator] = useState({ fn: mergeProps });
+  const [deduplicationProps, setDeduplicationProps] = useState(() => getPropsToMerge(appLayoutProps, appLayoutState));
+
+  const { registered, toolbarProps } = useMultiAppLayout(
+    forceDeduplicationType,
+    appLayoutState.isIntersecting,
+    deduplicationProps,
+    deduplicator.fn
+  );
+
+  useLayoutEffect(() => {
+    stateManager.current.setState = (appLayoutState, skeletonAttributes, deduplicationProps, mergeProps) => {
+      setAppLayoutState(appLayoutState);
+      setSkeletonAttributes(skeletonAttributes);
+      setDeduplicationProps(deduplicationProps);
+      setDeduplicator({ fn: mergeProps });
     };
```

Comment:
Important place no. 1 – this is how we synchronise the local and async parts of this code. See `stateManager.current.setState` above to see where this is called